cli/lsp/tsc.rs

```diff
@@ -1812,8 +1817,36 @@ impl<'a> State<'a> {
       response: None,
       state_snapshot,
       snapshots: HashMap::default(),
+      specifiers: HashMap::default(),
     }
   }
+
+  /// If a normalized version of the specifier has been stored for tsc, this
+  /// will "restore" it for communicating back to the tsc language server,
+  /// otherwise it will just convert the specifier to a string.
+  fn denormalize_specifier(&self, specifier: &ModuleSpecifier) -> String {
+    let specifier_str = specifier.to_string();
+    self
+      .specifiers
+      .get(&specifier_str)
+      .unwrap_or(&specifier_str)
+      .to_string()
+  }
+
+  /// In certain situations, tsc can request "invalid" specifiers and this will
+  /// normalize and memoize the specifier.
+  fn normalize_specifier<S: AsRef<str>>(
+    &mut self,
+    specifier: S,
+  ) -> Result<ModuleSpecifier, AnyError> {
+    let specifier_str = specifier.as_ref().replace(".d.ts.d.ts", ".d.ts");
```

Comment:
TypeScript being "broken" is a matter of opinion, I highly doubt they would consider fixing something that works for what it was designed for.  TypeScript is hardcoded to "guess" at extensions, so us being explicit about extensions "breaks" TypeScript in a few different ways.  The problem is that TypeScript isn't using the "resolve module" here, or any "trapable" API.  It has an algorithm which mimics Node.js resolution to try to find the type definitions, and I think the 4 or 5 item it attempts is `${cwd}/${specifier}.d.ts` (it tries things like `${cwd}/node_modules/${specifier}/package.json` and `${cwd}/node_modules/${specifier}/index.d.ts` first and a few others).  Since we can't control the resolution algorithm in this case, we are stuck this matching on something that always gets `.d.ts` banged onto the end of it.

Because the logic could change, or be different, or whatever in the future is why I broke it out into methods to abstract the rest of the code from all of that.  If we find error in this, we should only need to change it in one spot.